File: 122.买卖股票的最佳时机-ii.cpp
/*
 * @lc app=leetcode.cn id=122 lang=cpp
 *
 * [122] 买卖股票的最佳时机 II
 */

#include <vector>

// @lc code=start
class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        if(prices.size() <= 1) return 0;
        int profit = 0;
        for(int i = 1; i < prices.size(); i++){
            if(prices[i] > prices[i-1]){
                profit += prices[i] - prices[i-1];
            }
        }
        return profit;
    }
};
// @lc code=end

File: 122.买卖股票的最佳时机-ii.rs
/*
 * @lc app=leetcode.cn id=122 lang=rust
 *
 * [122] 买卖股票的最佳时机 II
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut prices_stack = vec![];
        let mut result = 0;
        for price in prices {
            match prices_stack.len() {
                0 => { prices_stack.push(price); },
                1 => {
                    if price > prices_stack[0] {
                        prices_stack.push(price);
                    } else {
                        prices_stack[0] = price;
                    }
                },
                2 => {
                    if price > prices_stack[1] {
                        prices_stack[1] = price;
                    } else {
                        result += prices_stack[1] - prices_stack[0];
                        prices_stack.pop();
                        prices_stack[0] = price;
                    }
                },
                _ => {}
            }
        }
        if prices_stack.len() == 2 {
            result += prices_stack[1] - prices_stack[0];
        }
        return result;
    }
}
// @lc code=end

